docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,281 @@
+   For an example of Kubernetes secrets volumes, see the
+   [Kubernetes secrets documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+   Secrets distributed in this manner become part of the `etcd` database,
+   and thus it becomes necessary to specify a
+   [KMS provider for data encryption](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/)
+   to prevent `etcd` from storing plaintext versions of secrets.
+
+   * Docker:
+   Use of a RAM-based temporary file system is desirable to protect secrets,
+   but only Linux platforms support `tmpfs` volumes for secret sharing:
+   generally, `/run` on a Linux system is a `tmpfs` volume.
+   For reasons of portability, `/tmp/edgex/secrets` has been chosen as a
+   cross-platform alternative to `/run`:
+   on the MacOS Docker implementation,
+   `/tmp` is a white-listed shared host directory,
+   while `/run` is not.
+   For production hardening, `/tmp` should be replaced by `/run`
+   where it appears in `docker-compose` files.
+   If the container runtime in use does not support host fileystem bind mounts,
+   then a separate Docker secrets volume should be created for each service.
+
+   * Snaps:
+   Note that due to an existing [bug](https://bugs.launchpad.net/snap-confine/+bug/1620442),
+   `$XDG_RUNTIME_DIR` is set to a snap-specific path
+   (`/run/user/1000/snap.`_$SNAP_NAME_`/`),
+   but the directory is **neither created nor accessible to the snap**.
+   Because it may be necessary to share several secrets outside of the snap
+   --such as secret store tokens--
+   it is desirable to place them in a location that can be mapped to
+   a snap content interface, such as `$SNAP_DATA/secrets`.
+   Although not a `tmpfs`, this is the best available option at this time.
+   The snap bug should be escalated in priority.
+
+## Consequences
+
+As the existing implementation is not fully-compliant with this ADR,
+significant scope will be added to current and future EdgeX releases
+in order to bring the project into compliance.
+
+List of needed improvements:
+
+- PKI private keys
+  * All: Move to using Vault as system of origin for the PKI instead of the standalone `security-secrets-setup` utility.
+  * All: Cache the PKI for Consul and Vault on persistent disk; rotate occasionally.
+  * All: Investigate hardware protection of cached Consul and Vault PKI secret keys.
+  * Snaps: Relocate PKI into the secrets volume instead of its own folder.
+
+- Secret store master password
+  * All: Enable hooks for hardware protection of secret store master password.
+
+- Secret store per-service authentication tokens
+  * No changes required.
+
+- Postgres superuser password
+  * Randomize every boot and inject using environment variable injection.
+
+- MongoDB service account passwords
+  * No changes required.
+
+- Redis authentication password
+  * All: Move server password to configuration file on secrets volume.
```

Comment:
Should we also speak to re-factoring the existing implementation which relies on an external security-secret-store-read command called from the Redis startup script?

Reply:
Will discuss with @andresrinivasan.

Reply:
Currently `security-secretstore-read` is invoked as part of the `security-secretstore-setup` container and creates a file with the secret the Redis server should use for authentication. The Redis startup script reads that file and sets the secret.

Per [edgex-go #2636](https://github.com/edgexfoundry/edgex-go/issues/2636), I will be changing the strategy to directly set the secret in the Redis server. This means `security-secretstore-setup` will have a dependency on the Redis container. I believe this conforms to Distribution and consumption of secrets/Recommended practices/2

Reply:
> Per [edgex-go #2636](https://github.com/edgexfoundry/edgex-go/issues/2636), I will be changing the strategy to directly set the secret in the Redis server. This means `security-secretstore-setup` will have a dependency on the Redis container. I believe this conforms to Distribution and consumption of secrets/Recommended practices/2

@andresrinivasan  Hopefully this was meant to be the other way -- that redis requires the the secretstore-setup utility run first.

Reply:
@bnevis-i and I just chatted. There are two options

1. Override the Redis Dockefile command to start Redis and then immediately run the app currently-known-as `security-secretstore-read` to read the secret from vault and set it directly in Redis using `SET CONFIG REQUIREPASS` Redis command.

1. Keep `security-secretstore-read` where it is now but have it create a config file with the secret stored on the secure file system and start Redis with both its own config file and the generated config file with the secret from the secure file system.

As we chatted, Bryon and I agreed that (1) felt better and was Snap friendly with sequential commands (@tonyespy). Furthermore, the last time I checked, having multiple config files while technically supported had some issue and I received glares from my Redis Labs colleagues when I demonstrated using this. I will check if this is still not recommended.

Reply:
One other potential option would be to implement a Redis plugin for the Vault secrets engine which already supports many different databases:

https://www.vaultproject.io/docs/secrets/databases

That said, I'm fine with (1) as a short-term approach.

Reply:
Vault database secrets plugins assume that the database has a persistent superuser account, such that Vault can use it to provision user accounts.  The Redis passwords is not persistent, nor are multiple users supported.  I don't think it's a good fit.

Reply:
Redis 6 supports [both multiple users and persistent ACLs](https://redis.io/topics/acl). As part of reworking the secret injection, I was also going to upgrade EdgeX to Redis 6.

Reply:
> Redis 6 supports [both multiple users and persistent ACLs](https://redis.io/topics/acl). I hadn't thought of the vault plugin when I was thinking about how the secret injection would work. I was also going to upgrade EdgeX to Redis 6 as part of this exercise anyway; is there a preference to use the vault plugin (did I mention I haven't thought about that at all and I don't know what it entails).

I think it would be a great value add to have a Redis6 plugin for Vault, preferrably built-in, as we don't extend Vault in any way currently.  As far as turning to a database secrets engine to create accounts -- that should be covered in the bootstrapping ADR.

Reply:
I added a separate treatment for Redis5 and Redis6.